src/symbolic_dimensions.jl

```diff
@@ -35,49 +34,54 @@ to one which uses `Dimensions` as its dimensions (i.e., base SI units)
 `expand_units`.
 """
 struct SymbolicDimensions{R} <: AbstractDimensions{R}
-    _data::SA.SparseVector{R}
-
-    SymbolicDimensions(data::SA.SparseVector) = new{eltype(data)}(data)
-    SymbolicDimensions{_R}(data::SA.SparseVector) where {_R} = new{_R}(data)
+    nzdims::Vector{UInt8}
+    nzvals::Vector{R}
 end
 
 static_fieldnames(::Type{<:SymbolicDimensions}) = ALL_SYMBOLS
-data(d::SymbolicDimensions) = getfield(d, :_data)
-Base.getproperty(d::SymbolicDimensions{R}, s::Symbol) where {R} = data(d)[ALL_MAPPING[s]]
-Base.getindex(d::SymbolicDimensions{R}, k::Symbol) where {R} = getproperty(d, k)
+function Base.getproperty(d::SymbolicDimensions{R}, s::Symbol) where {R}
+    nzdims = getfield(d, :nzdims)
+    i = ALL_MAPPING[s]
+    ii = searchsortedfirst(nzdims, i)
+    (ii <= length(nzdims) && nzdims[ii] == i) ? getfield(d, :nzvals)[ii] : zero(R)
+end
+Base.propertynames(::SymbolicDimensions) = ALL_SYMBOLS
+Base.getindex(d::SymbolicDimensions, k::Symbol) = getproperty(d, k)
 constructor_of(::Type{<:SymbolicDimensions}) = SymbolicDimensions
 
-SymbolicDimensions{R}(d::SymbolicDimensions) where {R} = SymbolicDimensions{R}(data(d))
-(::Type{D})(; kws...) where {D<:SymbolicDimensions} = D(DEFAULT_DIM_BASE_TYPE; kws...)
-(::Type{D})(::Type{R}; kws...) where {R,D<:SymbolicDimensions} =
-    let constructor=constructor_of(D){R}
-        length(kws) == 0 && return constructor(SA.spzeros(R, length(ALL_SYMBOLS)))
-        I = [ALL_MAPPING[s] for s in keys(kws)]
-        V = [tryrationalize(R, v) for v in values(kws)]
-        data = SA.sparsevec(I, V, length(ALL_SYMBOLS))
-        return constructor(data)
+SymbolicDimensions{R}(d::SymbolicDimensions) where {R} = SymbolicDimensions{R}(getfield(d, :nzdims), convert(Vector{R}, getfield(d, :nzvals)))
+SymbolicDimensions(; kws...) = SymbolicDimensions{DEFAULT_DIM_BASE_TYPE}(; kws...)
+function SymbolicDimensions{R}(; kws...) where {R}
+    if isempty(kws)
+        return SymbolicDimensions{R}(Vector{UInt8}(undef, 0), Vector{R}(undef, 0))
     end
+    I = UInt8[ALL_MAPPING[s] for s in keys(kws)]
+    p = sortperm(I)
+    V = R[tryrationalize(R, kws[i]) for i in p]
+    return SymbolicDimensions{R}(permute!(I, p), V)
```

Comment:
```suggestion
    return SymbolicDimensions{R}(I[p], V)
```

(The number of keywords is typically ~1-3 elements so `permute!` is a bit overkill)

Reply:
Surely a micro-optimization but still saves a few allocations 🤷 Note that the computer will also allocate more memory than needed (so not only for 1-3 elements) to avoid that new memory has to be allocated for every `push!`. (With the implementation on the master branch, `I` always contains 7 elements when converting `Dimensions` to `SymbolicDimensions` but with this PR zero dimensions will be omitted.)

Reply:
Okay, fine with me!